### PR TITLE
Logging: Add customizable logging backends and fmtlib based macros

### PR DESCRIPTION
**Pull request by daniellimws**
Continue #3449 

* Fix compile error
* Fix logging to file not working

Pushing this first for review, now looking into the problem with RemoveBackend.

---
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/3496)

### CODE REVIEW

File: src/common/logging/backend.h [hunk 1]
```diff
@@ -22,15 +26,93 @@ struct Entry {
...
+/**
+ * Backend that writes to a file passed into the constructor. If a log level is error or higher, it
+ * will flush immediately after writing
+ */
+class FileBackend : public Backend {
+public:
+    explicit FileBackend(const std::string& filename);
+
+    static const char* Name() {
+        return "console";
```
Comment (valentinvanelslande) [inline src/common/logging/backend.h]: file

Comment (daniellimws) [reply src/common/logging/backend.h]: hmm oh lol oops

**Comment by jroweboy** [comment]
My apologies! We just pushed a clang format update, and it looks like you will need to rebase this PR off the latest master and fix any conflicts. With the new clang format update, its really easy for people to fix any clang format issues, simply run the clang format build target. See the contributors guide for more information or ask me on IRC or discord. Thanks!

File: src/common/logging/backend.cpp [hunk 2]
```diff
@@ -6,7 +6,6 @@
 #include <array>
 #include <chrono>
 #include <cstdio>
-#include <future>
```
Comment (lioncash) [inline src/common/logging/backend.cpp]: <mutex> and <condition_variable> need to be included.

File: src/citra_qt/debugger/console.cpp [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+#include "citra_qt/debugger/console.h"
+#include "citra_qt/ui_settings.h"
+#include "common/logging/backend.h"
+
+namespace Debugger {
+void ToggleConsole() {
+#ifdef _WIN32
+    if (UISettings::values.show_console) {
+        if (AllocConsole()) {
+            freopen_s((FILE**)stdin, "CONIN$", "r", stdin);
```
Comment (lioncash) [inline src/citra_qt/debugger/console.cpp]: Why do casting like this when `&stdin` can just be done instead?

File: src/common/threadsafe_queue.h [hunk 4]
```diff
@@ -75,6 +83,25 @@ class SPSCQueue {
...
     }
 
+    /**
+     * Waits up to timeout for data to be Pushed to the queue. Push uses a condition variable to
+     * signal the waiting thread, but only if NeedSize = true. Returns false if the timeout is
+     * triggered. If the condition variable is signalled, returns the value from Pop
+     * @param T In parameter to store the value if this method returns true
+     * @param timeout Time in milliseconds to wait for a signal from a Push
+     */
+    bool PopWait(T& t, u64 timeout = 500) {
```
Comment (lioncash) [inline src/common/threadsafe_queue.h]: This should use `std::chrono::milliseconds` as type of the `timeout` argument itself. It enforces the API at callsites.

File: src/common/threadsafe_queue.h [hunk 5]
```diff
@@ -9,6 +9,7 @@
 
 #include <algorithm>
 #include <atomic>
+#include <chrono>
```
Comment (lioncash) [inline src/common/threadsafe_queue.h]: <condition_variable> should also be included.

**Comment by daniellimws** [comment]
@lioncash So should I close this and open a new PR?

**Comment by lioncash** [comment]
As you've squashed the changes into one commit, unfortunately yes.

In the future, you can put the changes on a different branch (just doing `git checkout -b NameOfNewBranch` while on this branch, then switching back to the original branch with git checkout and resetting it back to relevant changes.

Once the macros are in, it gets all the fmt stuff in place for others to use them with the logger (I don't know why a new set of macros were created instead of just re-adapting the existing ones, but I digress). Then a PR doing the log configuration stuff can be done separately.

**Comment by daniellimws** [comment]
Anyways feels like something broke in the midst of the rebase, perhaps easier to just start over :confused:

@jroweboy anything to say about this before I close it?

**Comment by jroweboy** [comment]
> I disagree with bundling both the libfmt-related changes and the changes to customization in the same PR. Land the formatting macros first. Then PR the other changes. One PR, one central change should be the model here.

wwylele commented something similar, so it sounds fine to me as well. go ahead and split it up.

The idea behind including each change was to be a proof of concept that the backends functioned, and i went ahead and included it as it was small and not too distracting.

> So should I close this and open a new PR?

if you struggle with git, thats the easiest course of action. you kinda messed up the rebase it looks like and re applied the merged commit on top of the previous commits or something :) Notice how it says you have 19 commits instead of the 1 that should be expected after a rebase?

> Once the macros are in, it gets all the fmt stuff in place for others to use them with the logger (I don't know why a new set of macros were created instead of just re-adapting the existing ones, but I digress). Then a PR doing the log configuration stuff can be done separately.

I addressed this in the original PR. The idea is to migrate while disturbing as little of the existing PRs as possible. Since its going to take a little bit of time to migrate all the old logging statements to the new one, we decided that we should make a new logging macro, migrate to the new logging macro, then do one quick rename to the old logging macro name. Voila all logging is just like it was before but with fmtlib now :)

Last words: good work so far. sorry if i wasn't clear that the work is like 80% done when i said its a decent task that should be picked up.

**Comment by daniellimws** [comment]
Ok closing this and opening separate prs.

**Pull request closed without merging**